packages/website/src/docs/extending-cadl/emitters-basics.md

```diff
@@ -0,0 +1,155 @@
+  const outputPath = Path.join(program.compilerOptions.outputPath!, "hello.txt");
+  const name = options.targetName;
+  await program.host.writeFile(outputPath, `hello ${name}!`);
+}
+```
+
+## Querying the program
+
+One of the main tasks of an emitter is finding types to emit. There are two main approaches: using the Semantic Walker, which lets you easily run code for every type in the program, and doing a custom traversal, which gives you a lot more flexibility.
+
+### Semantic Walker
+
+The Semantic Walker will visit every type in the Cadl program and call any callbacks you provide for that type. To use, import `navigateProgram` from `@cadl-lang/compiler`. Starting a walk needs two parameters - the program to walk, and an object with callbacks for each type. For example, if we want to do something for every model in the program, we could do the following in our `$onEmit` function:
+
+```typescript
+navigateProgram(program, {
+  model(m) {
+    // emit m
+  },
+});
+```
+
+You can provide a callback for every kind of Cadl type. The walker will call your callback pre-order, i.e. as soon as it sees a type for the first time it will invoke your callback. You can invoke callback post-order instead by prefixing the type name with `exit`, for example `exitModel(m)`.
+
+Note that the semantic walker will visit all types in the program including built-in Cadl types and cadl types defined by any libraries you're using. Care must be taken to filter out any types you do not intend to emit. Sometimes this is quite difficult, so a custom traversal may be easier.
+
+### Custom traversal
+
+Often times you will want to emit specific types, for example types that have a particular decorator or are in a particular namespace. In such cases it is often easier to write a custom traversal to find and emit those types. Once you have a type, you can access its [various fields](#todo) to and emit those types as well if needed.
+
+For example, let's say we want to emit a text file of model names but only if it has an `@emitThis` decorator. We could filter out such models in the Semantic Walker `model` callback, but it is more efficient to implement `@emitThis` such that it keeps a list of all the types its attached to and iterate that list. We can then traverse into types it references if needed.
+
+The following example will emit models with the `@emitThis` decorator and also any models referenced by that model.
+
+```typescript
+import {
+  DecoratorContext,
+  Model,
+  createStateSymbol,
+  createDecoratorDefinition,
+} from "@cadl-lang/compiler";
+
+// Decorator Setup Code
+const emitThisDef = createDecoratorDefinition({
+  name: "@emitThis",
+  target: ["Model"],
+  args: [],
+} as const);
+
+const emitModels = new WeakMap<Program, Type[]>();
+
+// @emitThis decorator
+export function $emitThis(context: DecoratorContext, target: Model) {
+  if (!emitThisDef.validate(context, target)) {
+    return;
+  }
+  let models = emitModels.get(context.program);
```

Comment:
Shouldn't we recommend using `program.stateSet` for that?

Reply:
Hmm yeah I suppose that is better.